Firestore/src/BulkWriter.php

```diff
@@ -0,0 +1,1226 @@
+            $path = $sentinel->fieldPath()
+            ? $sentinel->fieldPath()->pathString()
+            : null;
+
+            if (!$sentinel->includeInUpdateMask()) {
+                $excluded[] = $path;
+                continue;
+            }
+
+            $out[] = $path;
+        }
+
+        foreach ($paths as $path) {
+            $path = $path->pathString();
+            if (!in_array($path, $excluded)) {
+                $out[] = $path;
+            }
+        }
+
+        // The flag isn't really necessary since prefixes can only happen in
+        // update-paths, but using it lets us bypass the unnecessary check
+        // unless we need it.
+        if ($checkPrefixes) {
+            $this->checkPrefixes($out);
+        }
+
+        // Remove any duplicate values before returning.
+        return array_unique($out);
+    }
+
+    /**
+     * Check list of FieldPaths for prefix paths and throw exception.
+     *
+     * @param string[] $paths
+     * @throws \InvalidArgumentException If prefix paths are found.
+     */
+    private function checkPrefixes(array $paths)
+    {
+        sort($paths);
+
+        for ($i = 1; $i < count($paths); $i++) {
+            $prefix = $paths[$i - 1];
+            $suffix = $paths[$i];
+
+            $prefix = explode('.', $prefix);
+            $suffix = explode('.', $suffix);
+
+            $isPrefix = count($prefix) < count($suffix)
+            && $prefix === array_slice($suffix, 0, count($prefix));
+
+            if ($isPrefix) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Field path conflict detected for field path `%s`. ' .
+                    'Conflicts occur when a field path descends from another ' .
+                    'path. For instance `a.b` is not allowed when `a` is also ' .
+                    'provided.',
+                    implode('.', $prefix)
+                ));
+            }
+        }
+    }
+
+    /**
+     * Check whether BulkWriter is writeble with provided document.
+     *
+     * @param DocumentReference|string $document The document to target.
+     * @throws \InvalidArgumentException If document is not unique.
+     */
+    private function checkWriterConditions($document)
+    {
+        if ($this->closed) {
+            throw new \InvalidArgumentException('firestore: BulkWriter has been closed');
+        }
+        if ($this->transaction) {
+            throw new \InvalidArgumentException('firestore: BulkWriter has been initialized with transactionId');
+        }
+
+        if (is_null($document)) {
+            throw new \InvalidArgumentException('firestore: nil document contents');
+        }
+        if ($document instanceof DocumentReference) {
+            $document = $document->name();
+        }
+        if (in_array($document, $this->unique_documents)) {
+            throw new \InvalidArgumentException('firestore: bulkwriter: received duplicate mutations for path: ' . $document);
+        }
```

Comment:
Please note that `WriteBatch` will not be used to batch duplicate mutations on the same document.